### DISCUSSION TITLE
Isolated execution across classes, concurrent within classes

### DISCUSSION THREAD

**Discussion by martint** [junit-team/junit-framework#3526] [Q&A]
When running with these settings:

```
junit.jupiter.execution.parallel.enabled=true
junit.jupiter.execution.parallel.mode.default=concurrent
junit.jupiter.execution.parallel.mode.classes.default=concurrent
```

Is there a way to selectively isolate *classes* from each other while running all the tests in one class concurrently?

That is, given the following two test classes and methods:

```java
class Test1
{
    @Test
    public void test1a() {...}

    @Test
    public void test1b() {...}
}

class Test2
{
    @Test
    public void test2a() {...}

    @Test
    public void test2b() {...}
}
```

I'd like to be able to get the following interleaving, but only for these two classes. I'd like all other classes to be able to execute concurrently with each other.

```
<-- Test1.test1a() -->
<-- Test1.test1b() -->
                      <-- Test2.test2a() -->
                      <-- Test2.test2b() -->
```

I tried annotating the classes with `@ResourceLock` and `@Isolated`, but those seem to propagate to each individual test method and causes the tests within each classes to serialize with respect to each other.

**Comment by kriegaex** [comment]
I find it regrettable that this question got no attention for almost 14 months. FYI, there is a [related question on Stack Overflow](https://stackoverflow.com/q/79427865/1082681), even with a bounty, and nobody (including me) seems to know an answer to this arguably valid use case. In a nutshell: Use `@Isolated` on a test class, but run the methods inside that class concurrently, e.g. because the tests are slow and really do not need per-method isolation.

This discussion could in the end e.g. lead to an additional parameter for `@Isolated` by which to set the intra-class execution mode for sequential or concurrent. A related config key for _junit-platform.properties_ would also be advisable.

In Spock this can be configured, AFAIR. It would be nice to also have it in Jupiter. @marcphilipp, you know both tools and might want to say something about it. I apologise for naming you directly, but I would like this issue to gain some traction.

**Reply by mpkorstanje** [reply]
To wit Spock and JUnit Jupiter use the same mechanism for parallel execution, provided by the JUnit Platform. So I would be surprised if Spock was somehow more capable.

As for the reason this is getting little traction. I suspect that this due to a lack of a clear problem statement. In essence, what problem does the requested execution mode solve?

And though not explicitly mentioned anywhere, there is an issue template for feature request. So I would infer from that that feature request are to be made in an issue, not in the QA forum.

**Reply by marcphilipp** [reply]
Unfortunately, we don't always manage to find the time to respond to Q&A issues ourselves.

> This discussion could in the end e.g. lead to an additional parameter for `@Isolated` by which to set the intra-class execution mode for sequential or concurrent. A related config key for _junit-platform.properties_ would also be advisable.

I'd be open to discuss that. Feel free to submit a feature request!

> In Spock this can be configured, AFAIR. It would be nice to also have it in Jupiter. @marcphilipp, you know both tools and might want to say something about it. I apologise for naming you directly, but I would like this issue to gain some traction.

No worries! Since both Spock and Jupiter use the same mechanism under the hood, I doubt that this is configurable in Spock: 

https://github.com/junit-team/junit5/blob/ee52c64336ad3a079d598bd1c3884bafbc285047/junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/NodeTreeWalker.java#L81-L89

**Comment by kriegaex** [comment]
I know that there is a template for a feature request. But I searched first and found no issue, only this discussion. I thought it more prudent to continue the discussion first to find out if
* maybe there is another way to achieve this by means unknown to me and other users, such as to override a default/template class, implement a certain interface, add an extension or whatever,
* if the dev team would be open to a feature request at all, should one be necessary.

Then, there is still the "Reference in new issue" menu option here to evolve this exchange into an actual issue. I thought, if nobody responded to this inquiry in such a long time, maybe the same would happen to a feature request if not properly prepared.

> In Spock this can be configured, AFAIR.

Like I said, AFAIR. I have not had this scenario in a while, so I misremembered and both @mpkorstanje and @marcphilipp are right. Thanks for the correction, forcing me to refresh my imperfect memory. In my older projects, I solved this kind of problem via workarounds like splitting classes or running different suites separately. My bad, I should have checked first, but merely typed on my smartphone while riding a train yesterday. The Spock manual documents the actual situation [here](https://spockframework.org/spock/docs/2.4-M5/all_in_one.html#isolated-execution) (more explicitly than the [`@Isolated` javadoc](https://junit.org/junit5/docs/5.11.3/api/org.junit.jupiter.api/org/junit/jupiter/api/parallel/Isolated.html), BTW), just for reference.

The original problem statement might be a bit fuzzy, but yesterday I added some context, which is hopefully enough, especially if you also follow the Stack Overflow link where the author explains what he wants to achieve.
